cosmoz-page-location.html

```diff
@@ -164,23 +197,50 @@
 				_appHashChanged(_appHashString) {
 					const route = this._parse(_appHashString);
 					this._setHashBang(route.hashBang);
-					this.routePath = route.path;
-					this._microUpdate('routeHash', route.hash);
-					this._microUpdate('routeQuery', route.query);
+
+					if (this.setProperties) {
+						// Batch update of properties if using Polymer 2.x
+						this._batchMicroUpdate({
+							routePath: route.path,
+							routeHash: route.hash,
+							routeQuery: route.query
+						});
+					} else {
+						this.routePath = route.path;
+						this._microUpdate('routeHash', route.hash);
+						this._microUpdate('routeQuery', route.query);
+					}
 				},
 
 				_routeChanged(change) {
 					if (!this.isAttached) {
 						return;
 					}
-					const hash = this._encode(this.getRoute()),
-						path = change && change.path;
-					if (hash === this._appHashString) {
+
+					const path = change && change.path;
+
+					let route,
+						routeHashUri;
+
+					if (path && path.startsWith('routeHash')) {
+						// See issue #36
+						// If this cosmoz-page-location query parameters are not in sync with the URL, and we are trying to update
+						// some hash params, then we will update the URL with incorrect route hash string
+						// To avoid that, use the actual location
+						route = this._parse(window.decodeURIComponent(window.location.hash.slice(1)));
+						routeHashUri = this._encode({ hashBang: route.hashBang, path: route.path, query: route.query, hash: this.routeHash });
+					} else {
+						route = this.getRoute();
+						routeHashUri = this._encode(route);
+					}
+
+					if (this._appHashString === routeHashUri) {
 						return;
 					}
+
 					this.$.location.dwellTime = path && path.length ? this.microDwellTime : this.dwellTime ;
-					this._appHashString = hash;
-				}
+					this._appHashString = routeHashUri;
+				},
```

Comment:
Misplaced comma?

Reply:
Yes, looks like a comma I forgot to remove after removing some test code.